src/realm/object-store/impl/deep_change_checker.cpp

```diff
@@ -0,0 +1,345 @@
+                             depth + 1);
+        }
+
+        auto& target = *table.get_link_target(ColKey(link.col_key));
+        auto lvr = obj.get_linklist(ColKey(link.col_key));
+        return std::any_of(lvr.begin(), lvr.end(), [&, this](auto key) {
+            return this->check_row(target, key.value, filtered_columns, depth + 1);
+        });
+    };
+
+    // Check the `links` of all `m_related_tables` and return true if any of them has a `linked_object_changed`.
+    return std::any_of(begin(it->links), end(it->links), linked_object_changed);
+}
+
+bool DeepChangeChecker::check_row(Table const& table, ObjKeyType object_key, std::vector<ColKey> filtered_columns,
+                                  size_t depth)
+{
+    // Arbitrary upper limit on the maximum depth to search
+    if (depth >= m_current_path.size()) {
+        // Don't mark any of the intermediate rows checked along the path as
+        // not modified, as a search starting from them might hit a modification
+        for (size_t i = 0; i < m_current_path.size(); ++i)
+            m_current_path[i].depth_exceeded = true;
+        return false;
+    }
+
+    TableKey table_key = table.get_key();
+
+    // If the pair (table_key.value, key) can be found in `m_info.tables` we can
+    // end the search and return here.
+    if (depth > 0) {
+        auto it = m_info.tables.find(table_key.value);
+        if (it != m_info.tables.end() && it->second.modifications_contains(object_key, filtered_columns))
+            return true;
+    }
+
+    // Look up the unmodified objects for the `table_key.value` and check if the
+    // `key` can be found within them. If so, we can return without checking the
+    // outgoing links.
+    auto& not_modified = m_not_modified[table_key.value];
+    auto it = not_modified.find(object_key);
+    if (it != not_modified.end())
+        return false;
+
+    // If both of the above short cuts don't lead to a result we need to check the
+    // outgoing links.
+    bool ret = check_outgoing_links(table, object_key, filtered_columns, depth);
+    if (!ret && (depth == 0 || !m_current_path[depth - 1].depth_exceeded))
+        not_modified.insert(object_key);
+    return ret;
+}
+
+bool DeepChangeChecker::operator()(ObjKeyType key)
+{
+    // If the root object changed we do not need to iterate over every row since a notification needs to be sent
+    // anyway.
+    if (m_root_object_changes &&
+        m_root_object_changes->modifications_contains(key, m_filtered_columns_in_root_table)) {
+        return true;
+    }
+
+    return check_row(m_root_table, key, m_filtered_columns, 0);
+}
+
+KeyPathChangeChecker::KeyPathChangeChecker(TransactionChangeInfo const& info, Table const& root_table,
+                                           std::vector<RelatedTable> const& related_tables,
+                                           std::vector<KeyPathArray> key_path_arrays)
```

Comment:
If we're going to take a vector of vectors, can we either pas it by const ref or std::move() it in the constructor?